lsp/core/aws-lsp-core/src/content/handlers/cachedContentHandler.ts

```diff
@@ -0,0 +1,91 @@
+import { URI } from 'vscode-uri'
+import { DefaultHttpRequester, HttpRequester, getETag } from '../../http/requesters'
+import { TimeProvider } from '../../util/timeProvider'
+import { ContentMetadata, UriCacheRepository } from '../cache/uriCacheRepository'
+import {
+    ContentRequestMiddleware,
+    ContentRequestMiddlewareDelegate,
+    ContentRequestResponse,
+} from '../contentRequestMiddleware'
+
+export interface CachedContentHandlerProps {
+    cacheRepository: UriCacheRepository
+    httpRequester?: HttpRequester
+    timeProvider?: TimeProvider
+}
+
+/**
+ * Middleware that will use a cached version of content if the online version has not been updated.
+ * Newer content that has been downloaded will be cached.
+ *
+ * This avoids unnecessary downloads.
+ *
+ * Cached content is considered stale after {@link thirtyMinutesInMillis}
+ * and is refreshed.
+ */
+export class CachedContentHandler implements ContentRequestMiddleware {
+    private static readonly thirtyMinutesInMillis = 60 * 30 * 1000
+    public static readonly timeoutPeriodInMillis = CachedContentHandler.thirtyMinutesInMillis
+
+    private readonly cacheRepository: UriCacheRepository
+    private readonly httpRequester: HttpRequester
+    private readonly timeProvider: TimeProvider
+
+    constructor({
+        cacheRepository,
+        httpRequester = new DefaultHttpRequester(),
+        timeProvider = new TimeProvider(),
+    }: CachedContentHandlerProps) {
+        this.cacheRepository = cacheRepository
+        this.httpRequester = httpRequester
+        this.timeProvider = timeProvider
+    }
+
+    async get(uri: URI, next: ContentRequestMiddlewareDelegate): Promise<ContentRequestResponse> {
+        if (!uri.scheme.startsWith('http')) {
+            // we're only caching from online sources
+            return next(uri)
+        }
+
+        const cachedMetadata = await this.cacheRepository.getContentMetadata(uri)
+        const currentETag = await this.getETag(uri)
+
+        if (this.isCacheStale(cachedMetadata) === false && cachedMetadata!.eTag === currentETag) {
```

Comment:
Is something special happening with this bool rather than to use `!this.isCacheStale(cachedMetadata)`?

Reply:
You're correct, the intent is "this is not stale content (and the etag hasn't changed)"